app/domain/authentication/authn_azure/application_identity.rb

```diff
@@ -0,0 +1,118 @@
+module Authentication
+  module AuthnAzure
+
+    Log = LogMessages::Authentication
+    Err = Errors::Authentication::AuthnAzure
+    # Possible Errors Raised: MissingNamespaceConstraint, IllegalConstraintCombinations,
+    # ScopeNotSupported, InvalidHostId
+
+    # This class defines an application identity of a given conjur host.
+    # The constructor initializes an ApplicationIdentity object and validates that
+    # it is configured correctly.
+    # The difference between the validation in this constructor and
+    # the validation in ValidateApplicationIdentity is that here we validate that
+    # the application identity is configured correctly, and thus is a valid application
+    # identity. In ValidateApplicationIdentity we validate that the defined application
+    # identity is actually the correct one in kubernetes.
+    # For example, an application identity `some-namepsace/blah/some-value` is not
+    # valid and will fail the validation here. However, an application identity
+    # `some-namespace/service-account/some-value` is a valid application identity
+    # and will pass the validation here. If the host is actually running from
+    # a pod with service account `some-other-value` then it will fail the
+    # validation of ValidateApplicationIdentity
+    class ApplicationIdentity
+
+      def initialize(host_annotations:, service_id:)
+        @host_annotations = host_annotations
+        @service_id       = service_id
+
+        validate
+      end
+
+      # add validator to check resource group and subscription id are present
+      def constraints
+        @constraints ||= {
+            subscription_id:          constraint_value("subscription-id"),
+            resource_group:           constraint_value("resource-group"),
+            user_assigned_identity:   constraint_value("user-assigned-identity"),
+            system_assigned_identity: constraint_value("system-assigned-identity"),
+        }.compact
+      end
+
+      private
+
+      # Validates that the application identity is defined correctly
+      def validate
+        # Check if annotations present are part of the permitted annotations for authn-azure
+        validate_permitted_annotations
+
+        # validate that a constraint exists on the required annotations
+        raise Err::MissingAnnotationConstraint.new(constraints.subscription_id) unless constraint_from_annotation constraints.has_key? "subscription_id"
```

Comment:
> The error message is incorrect. It will print the value of the annotation instead of the annotation name.

The actual error return `MissingAnnotationConstraint` is incorrect? or the `constraints.subscription_id` is incorrect?

Reply:
`constraints.subscription_id` is the annotation value

Reply:
From my understanding, when using `unless` you need something that will return T or F so what about
```
raise Err::MissingAnnotationConstraint.new(constraints.subscription_id) unless !(constraint_from_annotation"subscription_id").nil?
```
To check if the annotation has for example `authn-azure/subscription_id`

BTW, will update the first part according to your response to my previous comment